### PR TITLE
feat: deterministic workflow rules via `.specs/instructions.md`

### PR DESCRIPTION
**Pull request by hiromaily**
# feat: deterministic workflow rules via `.specs/instructions.md`

## Summary

Adds an optional per-repository `.specs/instructions.md` file that lets maintainers declare **deterministic workflow rules** which the Go engine enforces at phase-4 completion. When task-decomposer produces a `tasks.md` that matches a rule's `when` block but is missing `mode: human_gate`, the engine automatically triggers a REVISE loop — no LLM judgement involved in enforcement.
This makes the existing (but previously dormant) `human_gate` execution mode fire automatically based on repository-specific conditions, instead of relying on task-decomposer's generic prompt guidance to spot them.

## Motivation

`[human_gate]` / `mode: human_gate` has been fully wired into the engine, parser, state, and orchestrator for some time — but in practice it never fires. The reason: `agents/task-decomposer.md` only lists generic criteria ("merging a PR in another repository, updating an external dependency…") with no way to express repository-specific rules.
Concrete example: in `main-repo`, any task touching `**/*.proto` files needs coordination with the external `sub-proto` repository before it can proceed. task-decomposer has no way to learn this, so such tasks are generated as regular `mode: sequential` implementation tasks and the pipeline runs headlong into a broken build.
The plugin's own design principle (`docs/architecture/technical-decisions.md`) states:
> Prefer deterministic enforcement over prompt instructions.
This PR applies that principle to the cross-repo dependency problem.

### Rule file format

`.specs/instructions.md` at the repository root. Markdown with a single YAML frontmatter block; the markdown body is ignored by the validator (reserved for human notes).
```markdown---rules: - id: sub-app when: files_match: - "backend/**/*.proto" - "backend/gen/proto/**" require: human_gate reason: "merge PR on sub-app repository"
 - id: destructive-migration when: files_match: - "backend/migrations/**/*.sql" title_matches: "(?i)drop\\s+(table|column)" require: human_gate reason: "Stakeholder verification is required for this destructive migration."---```
**Schema:**
| Field | Type | Required | Description ||---|---|---|---|| `id` | string | yes | Unique identifier used in error messages || `when.files_match` | list of glob | at least one of files_match / title_matches | Doublestar glob patterns matched against each task's `files:` list || `when.title_matches` | Go regex | see above | Regex matched against the task title || `require` | string | yes | Must be `human_gate` (MVP) || `reason` | string | yes | Shown in the human_gate prompt when the task is reached |
- `when.files_match` is OR across patterns (any glob matching any file is a hit)- `files_match` and `title_matches` are ANDed when both are present- Globs use `doublestar` syntax (`**` supported); Go stdlib's `filepath.Match` is insufficient- Regex uses Go's `regexp` package (`(?i)` for case-insensitive)

### Evaluation flow

```phase-4 (task-decomposer) writes tasks.md ↓pipeline_report_result(phase=phase-4) ↓ParseTasksMd → validation.LoadRules → validation.Validate ↓ ┌─────┴─────┐ │ │ clean violations │ │ │ ├─► write review-tasks.md with REVISE verdict │ ├─► sm.RevisionBump(workspace, RevTypeTasks) │ └─► return next_action_hint="revision_required" │ ↓ │ handlePhaseFour next tick: │ ↓ │ review-tasks.md exists? │ ├─ TaskRevisions >= MaxRevisionRetries │ │ → checkpoint "task-workflow-rules-retry-limit" │ └─ otherwise │ → respawn task-decomposer with │ [ArtifactDesign, ArtifactReviewTasks] │ └─► stale review-tasks.md removed (if any) phase-4 PhaseComplete → advance to phase-4b normally```

### Architectural decisions

**Principle: each phase owns its own completion gate.** phase-4 detects the violation, phase-4 writes the REVISE signal, phase-4 re-runs. We do NOT fake `PhaseComplete` to reuse phase-4b's machinery — that would lie about phase state and create implicit coupling.
**Retry-limit enforcement is centralised in the engine.** `applyWorkflowRules` (in `tools/`) bumps `TaskRevisions` just like phase-4b does at `pipeline_report_result.go:188`. The enforcement check lives in `handlePhaseFour` (mirroring `handlePhaseFourB:343-349`). Bump in one place, check in one place — symmetric with the existing phase-4b REVISE path.
**The checkpoint action name is distinct.** `"task-workflow-rules-retry-limit"` vs phase-4b's `"task-retry-limit"` — so analytics and logs can tell the two escalation sources apart.
**Write-then-bump ordering.** On a violation: write `review-tasks.md` first, then `RevisionBump`. If the bump fails we lose one retry tick but stay correct. The reverse would silently waste a retry slot with no findings on disk.
**Fail-open on absence.** If `.specs/instructions.md` does not exist, `LoadRules` returns `(&WorkflowRules{}, nil)`. Zero behavioural change for repositories that don't opt in.
**Fail-loud on malformation.** YAML parse errors, unknown fields, invalid regex, invalid globs, unsupported `require:` values, missing required fields — all surface as hard errors at load time with line/field context.

### Non-goals (explicitly out of scope)

- **Tier 2 (natural language) rules interpreted by LLM agents** — rejected during design. Rules that cannot be mechanised should go in `CLAUDE.md` / `AGENTS.md` / `.kiro/steering/`, not here.- **Reading source file content** — only task `files:` list and title are inspected.- **Rule application in phases other than phase-4** — design, implementation, and review phases are untouched.- **`require:` values other than `human_gate`** — keeps MVP scope tight.- **User-level / global instruction files** — repository-only.

## Files changed

### New

| Path | Purpose ||---|---|| `mcp-server/internal/validation/workflow_rules.go` | Loader, glob matcher, regex matcher, validator, formatter || `mcp-server/internal/validation/workflow_rules_test.go` | External-package tests || `mcp-server/internal/validation/workflow_rules_internal_test.go` | Internal-package tests for private helpers || `mcp-server/internal/validation/testdata/workflow_rules/*.md` | 10 fixtures (happy path + 9 error/edge cases) || `docs/reference/workflow-instructions.md` | User-facing schema + evaluation flow + failure modes || `docs/ja/reference/workflow-instructions.md` | Japanese mirror |

### Modified

| Path | Change ||---|---|| `mcp-server/go.mod` + `go.sum` | Added `[github.com/bmatcuk/doublestar/v4`](http://github.com/bmatcuk/doublestar/v4) and `[gopkg.in/yaml.v3`](http://gopkg.in/yaml.v3) as direct dependencies || `mcp-server/internal/tools/pipeline_report_result.go` | New `applyWorkflowRules` helper + phase-4 branch in `determineTransition`; bumps `TaskRevisions` on violations; clears stale `review-tasks.md` on pass-through || `mcp-server/internal/tools/pipeline_report_result_test.go` | Added integration tests: violation, no-instructions-file, malformed instructions, rules-present-no-violations, flat-workspace fall-through, `TaskRevisions` monotonic increment; extracted `writeProtoRuleViolationFixture` / `setupPhase4SpecWorkspace` helpers || `mcp-server/internal/orchestrator/engine.go` | `handlePhaseFour` now dispatches on `review-tasks.md` presence: fresh run vs revision run, with `MaxRevisionRetries` enforcement || `mcp-server/internal/orchestrator/engine_test.go` | New table rows + standalone tests for `handlePhaseFour` fresh/revision/retry-limit branches || `agents/task-decomposer.md` | Informational note: "If `.specs/instructions.md` exists, its deterministic rules are enforced post-generation" || `README.md` | New "Repository workflow rules" section with the two canonical examples || `docs/architecture/pipeline-sequence.md` | Updated phase-4 mermaid diagram to include the workflow-rules branch || `docs/.vitepress/config.ts` | Added `workflow-instructions` to both English and Japanese sidebars |

### Explicitly NOT changed

- `SKILL.md` — orchestrator already handles `human_gate` actions- `mcp-server/internal/state/` schema — `TaskRevisions` and `MaxRevisionRetries` already exist- `mcp-server/internal/orchestrator/engine.go`'s `handlePhaseFourB` — existing phase-4b REVISE path is untouched- `mcp-server/internal/tools/task_ops.go` — `ParseTasksMd` unchanged- Hook scripts — no changes to hook behaviour

## Test plan

### Unit tests (`internal/validation/`)

- LoadRules happy path, missing file, empty rules list, 7 parse error cases (unknown field, bad regex, bad require, missing id, no frontmatter, bad glob, no conditions)- matchFiles: recursive proto, first-win ordering, no match, empty files- matchTitle: match upper/lower, with prefix, nil regex- Validate: files-only rule, title-only rule, AND semantics (all 3 truth-table cells), skip tasks already in `human_gate`, multiple violations, sorted output, empty rules, nil rules, empty files list, `MatchedBy` format contract- FormatReviewFindings: violation rendering, empty input returns `""`

### Integration tests (`internal/tools/`)

- Phase-4 with proto-file violation → `revision_required` + `review-tasks.md` written + `TaskRevisions == 1` + phase-4 not in `CompletedPhases`- Phase-4 without `instructions.md` → proceeds normally- Phase-4 with malformed `instructions.md` → MCP error surfaced- Phase-4 with rules present but no violations → proceeds, stale `review-tasks.md` cleaned up- Phase-4 with flat workspace (not under `.specs/`) → falls through safely- Phase-4 with repeated violations → `TaskRevisions` monotonically increments

### Engine tests (`internal/orchestrator/`)

- `handlePhaseFour` fresh run: spawns task-decomposer with `[ArtifactDesign]` only- `handlePhaseFour` revision run: spawns task-decomposer with `[ArtifactDesign, ArtifactReviewTasks]` and revised prompt- `handlePhaseFour` retry limit: returns `NewCheckpointAction("task-workflow-rules-retry-limit", ..., ["approve", "abandon"])`

### Verification

- [ ] `cd mcp-server && go build ./...`- [ ] `cd mcp-server && go test -race ./...` — 14 packages green- [ ] `cd mcp-server && go tool golangci-lint run ./...` — 0 issues- [ ] `bash scripts/test-hooks.sh` — 62/62 pass

### Manual smoke test

1. Create a test repo with a `.specs/instructions.md` containing a proto-file rule2. Run `/forge` on a task that touches `**/*.proto`3. Observe: task-decomposer's first pass either emits `mode: human_gate` correctly (if it reads the instructions) or gets REVISE-looped by the Go validator4. On the revision run, task-decomposer receives `review-tasks.md` in its inputs and should fix the violation5. At `MaxRevisionRetries`, the pipeline should emit the `task-workflow-rules-retry-limit` checkpoint

## Migration / backward compatibility

- **100% backward compatible.** Repositories without `.specs/instructions.md` behave exactly as before — `LoadRules` returns an empty rule set and `Validate` short-circuits.- No state schema changes; existing `state.json` files work unchanged.- No changes to the orchestrator skill or hook scripts; existing pipelines run identically when no rules file is present.

## Follow-ups (not in this PR)

- **Rule application in other phases** — e.g. a design-phase rule that escalates on public API changes.- **Content-based matching** — reading source file content for regex matches, not just task metadata.- **Rule composition** — OR groups, priority, severity levels.- **MCP lint tool** — `mcp__forge-state__validate_rules` to lint `.specs/instructions.md` in isolation.- **Shared `triggerTaskRevision` helper** — extract the bump+limit-check pattern so phase-4 and phase-4b share one pathway (Option A' from the design discussion).

### CODE REVIEW

File: mcp-server/internal/tools/pipeline_report_result.go [hunk 1]
```diff
@@ -573,3 +586,110 @@ func clearCompletedFailTasks(sm *state.StateManager, workspace string) error {
...
+	return reportResultResponse{
+		StateUpdated:    true,
+		ArtifactWritten: artifactWritten,
+		VerdictParsed:   "REVISE",
+		Findings:        findings,
+		NextActionHint:  "revision_required",
+		Warning: fmt.Sprintf("phase-4 workflow rules: %d violation(s) — see %s",
+			len(violations), state.ArtifactReviewTasks),
+	}, true, nil
+}
```
[the remark this answers is not in the document]
Comment (hiromaily) [reply mcp-server/internal/tools/pipeline_report_result.go:716]: Extracted two helpers: `loadPhase4Context` (reads tasks.md, checks workspace layout, loads rules) and `writeViolationResponse` (writes review-tasks.md, bumps TaskRevisions, builds response). `applyWorkflowRules` now orchestrates the two helpers and the no-violation cleanup, keeping each function to a single responsibility.

File: mcp-server/internal/orchestrator/engine.go [hunk 2]
```diff
@@ -302,14 +302,52 @@ func (*Engine) handleCheckpointA(st *state.State) (Action, error) {
...
 		agentTaskDecomposer,
-		"Decompose the design into implementation tasks.",
+		"Revise task decomposition based on workflow-rules findings in review-tasks.md.",
 		state.DefaultModel,
 		PhaseFour,
-		[]string{state.ArtifactDesign},
+		[]string{state.ArtifactDesign, state.ArtifactReviewTasks},
 		state.ArtifactTasks,
 	), nil
 }
```
[the remark this answers is not in the document]
Comment (hiromaily) [reply mcp-server/internal/orchestrator/engine.go]: Extracted `phaseFourFreshAction()` (returns the initial spawn action) and `phaseFourRevisionAction(st)` (handles retry-limit check + revision spawn) as private helpers. `handlePhaseFour` now just stats the review file and dispatches to the appropriate helper.
